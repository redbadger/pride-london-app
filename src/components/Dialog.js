// @flow
import React from "react";
import type { Node } from "react";
import {
  Modal,
  StyleSheet,
  TouchableWithoutFeedback,
  View
} from "react-native";
import Button from "./ButtonPrimary";
import {
  dialogBackdropColor,
  dialogBgColor,
  dialogHeaderDividerColor
} from "../constants/colors";

type Props = {
  applyButtonText: string,
  applyButtonLabel: string,
  applyButtonDisabled?: boolean,
  children: Node,
  headerLeft?: Node,
  headerRight?: Node,
  onApply: Function,
  onCancel?: Function,
  title: Node,
  visible: boolean
};

const Dialog = ({
  applyButtonText,
  applyButtonLabel,
  applyButtonDisabled,
  children,
  headerLeft,
  headerRight,
  onApply,
  onCancel,
  title,
  visible
}: Props) => (
  <Modal
    animationType="fade"
    onRequestClose={onCancel}
    transparent
    visible={visible}
  >
    <TouchableWithoutFeedback
      style={styles.backdrop}
      onPress={onCancel}
      accessible={false}
    >
      <View style={styles.container}>
        <TouchableWithoutFeedback accessible={false}>
          <View>
            <View style={styles.content}>
              <View style={styles.header}>
                <View style={styles.headerSideLeft}>{headerLeft}</View>
                <View style={styles.headerSideMiddle}>{title}</View>
                <View style={styles.headerSideRight}>{headerRight}</View>
              </View>
              {children}
            </View>
            <View style={styles.applyButtonContainer}>
              <Button
                disabled={applyButtonDisabled}
                onPress={onApply}
                accessibilityLabel={applyButtonLabel}
              >
                {applyButtonText}
              </Button>
            </View>
          </View>
        </TouchableWithoutFeedback>
      </View>
    </TouchableWithoutFeedback>
  </Modal>
);

Dialog.defaultProps = {
  headerLeft: undefined,
  headerRight: undefined,
  onCancel: () => {},
  applyButtonDisabled: false
};

const styles = StyleSheet.create({
  backdrop: {
    flex: 1
  },
  container: {
    flex: 1,
    justifyContent: "center",
    padding: 16,
    backgroundColor: dialogBackdropColor
  },
  content: {
    backgroundColor: dialogBgColor,
    borderRadius: 4,
    paddingBottom: 8
  },
  header: {
    flexDirection: "row",
    alignItems: "center",
    justifyContent: "space-between",
    height: 40,
    paddingHorizontal: 16,
    borderBottomWidth: 1,
    borderBottomColor: dialogHeaderDividerColor
  },
  headerSideLeft: {
    alignItems: "flex-start"
  },
  headerSideMiddle: {
    alignItems: "center"
  },
  headerSideRight: {
    alignItems: "flex-end"
  },
  applyButtonContainer: {
    marginTop: 8
  }
});

export default Dialog;
