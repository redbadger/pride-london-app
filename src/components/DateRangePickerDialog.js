// @flow
import React from "react";
import { StyleSheet, View } from "react-native";
import DateRangePicker from "./DateRangePicker";
import Dialog from "./Dialog";
import Text from "./Text";
import Touchable from "./Touchable";
import text from "../constants/text";
import type { DateRange } from "../data/date-time";
import { formatDateRange } from "../data/formatters";
import { dialogTitleColor } from "../constants/colors";

type Props = {
  applyButtonText: string,
  applyButtonLabel: string,
  applyButtonDisabled?: boolean,
  dateRange: ?DateRange,
  onApply: () => void,
  onCancel: () => void,
  onChange: (?DateRange) => void,
  visible: boolean,
  forceNewRange: boolean
};

const formatTitle = (dateRange: ?DateRange): string => {
  if (!dateRange) return text.filterDayPickerTitle;

  return (
    formatDateRange(dateRange) +
    (dateRange.startDate === dateRange.endDate ? " -" : "")
  );
};

const formatTitleLabel = (dateRange: ?DateRange): string => {
  if (!dateRange) return text.filterDayPickerTitle;

  return `Selected: ${formatDateRange(dateRange)} ${
    dateRange.startDate === dateRange.endDate
      ? ", pick another day to select range"
      : ""
  }`;
};

class DateRangePickerDialog extends React.PureComponent<Props> {
  static defaultProps = {
    applyButtonDisabled: false
  };

  clear = () => {
    this.props.onChange(null);
  };

  render() {
    const { dateRange, forceNewRange } = this.props;
    const title = formatTitle(dateRange);
    const titleLabel = formatTitleLabel(dateRange);

    return (
      <Dialog
        applyButtonText={this.props.applyButtonText}
        applyButtonLabel={this.props.applyButtonLabel}
        applyButtonDisabled={this.props.applyButtonDisabled}
        title={
          <Text
            type="h3"
            style={styles.headerTitle}
            accessible
            accessibilityLabel={titleLabel}
            accessibilityTraits={["header"]}
            accessibilityLiveRegion="polite"
          >
            {title}
          </Text>
        }
        headerRight={
          dateRange && (
            <Touchable
              style={styles.clearButton}
              onPress={this.clear}
              accessibilityLabel="Clear date selection"
            >
              <Text type="small" style={{ color: dialogTitleColor }}>
                Clear
              </Text>
            </Touchable>
          )
        }
        headerLeft={dateRange && <View style={styles.clearButton} />}
        onApply={this.props.onApply}
        onCancel={this.props.onCancel}
        visible={this.props.visible}
      >
        <DateRangePicker
          onChange={this.props.onChange}
          dateRange={dateRange}
          forceNewRange={forceNewRange}
        />
      </Dialog>
    );
  }
}

const styles = StyleSheet.create({
  clearButton: {
    width: 44
  },
  headerTitle: {
    color: dialogTitleColor,
    // Needs to start higher up on screen than the 'Clear' button for a11y order
    height: 40,
    paddingTop: 10
  }
});

export default DateRangePickerDialog;
