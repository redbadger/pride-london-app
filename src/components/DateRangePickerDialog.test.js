// @flow
import React from "react";
import { shallow } from "enzyme";
import DateRangePickerDialog from "./DateRangePickerDialog";
import Dialog from "./Dialog";
import type { DateRange } from "../data/date-time";

const render = props =>
  shallow(
    <DateRangePickerDialog
      applyButtonText="Apply"
      dateRange={(null: ?DateRange)}
      onApply={() => {}}
      onCancel={() => {}}
      onChange={() => {}}
      visible
      forceNewRange={false}
      {...props}
    />
  );

describe("renders correctly", () => {
  it("without date selection", () => {
    const output = render();
    expect(output).toMatchSnapshot();
  });

  it("with date selection", () => {
    const output = render({
      dateRange: { startDate: "2018-02-02", endDate: "2018-02-02" }
    });
    expect(output).toMatchSnapshot();
  });

  it("with range selection", () => {
    const output = render({
      dateRange: { startDate: "2018-02-02", endDate: "2018-02-08" }
    });
    expect(output).toMatchSnapshot();
  });
});

it("calls onChange with null when user presses clear button", () => {
  const onChange = jest.fn();
  const output = render({ onChange });
  const clearButton = output.find(Dialog).prop("headerRight");
  clearButton.props.onPress();

  expect(onChange).toHaveBeenCalledWith(null);
});
