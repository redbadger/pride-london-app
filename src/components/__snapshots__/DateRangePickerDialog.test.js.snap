// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly with date selection 1`] = `
<Dialog
  applyButtonLabel="Apply selection"
  applyButtonText="Apply"
  headerRight={
    <Touchable
      accessibilityComponentType="button"
      accessibilityLabel="Clear date selection"
      accessibilityTraits={
        Array [
          "button",
        ]
      }
      accessible={undefined}
      delayLongPress={undefined}
      delayPressIn={50}
      delayPressOut={undefined}
      disabled={undefined}
      hitSlop={undefined}
      onLongPress={undefined}
      onPress={[Function]}
      onPressIn={undefined}
      onPressOut={undefined}
      pressRetentionOffset={undefined}
      style={undefined}
    >
      <Text
        color="blackColor"
        markdown={false}
        markdownStyle={Object {}}
        style={
          Object {
            "color": "rgb(45, 47, 127)",
          }
        }
        type="small"
      >
        Clear
      </Text>
    </Touchable>
  }
  onApply={[Function]}
  onCancel={[Function]}
  title={
    <Text
      accessibilityLabel="Selected: 2 Feb , pick another day to select range"
      accessibilityLiveRegion="polite"
      accessibilityTraits={
        Array [
          "header",
        ]
      }
      accessible={true}
      color="blackColor"
      markdown={false}
      markdownStyle={Object {}}
      style={
        Object {
          "alignSelf": "flex-end",
          "color": "rgb(45, 47, 127)",
          "height": 40,
          "paddingTop": 10,
        }
      }
      type="h3"
    >
      2 Feb -
    </Text>
  }
  visible={true}
>
  <DateRangePicker
    dateRange={
      Object {
        "endDate": "2018-02-02",
        "startDate": "2018-02-02",
      }
    }
    forceNewRange={false}
    onChange={[Function]}
  />
</Dialog>
`;

exports[`renders correctly with range selection 1`] = `
<Dialog
  applyButtonLabel="Apply selection"
  applyButtonText="Apply"
  headerRight={
    <Touchable
      accessibilityComponentType="button"
      accessibilityLabel="Clear date selection"
      accessibilityTraits={
        Array [
          "button",
        ]
      }
      accessible={undefined}
      delayLongPress={undefined}
      delayPressIn={50}
      delayPressOut={undefined}
      disabled={undefined}
      hitSlop={undefined}
      onLongPress={undefined}
      onPress={[Function]}
      onPressIn={undefined}
      onPressOut={undefined}
      pressRetentionOffset={undefined}
      style={undefined}
    >
      <Text
        color="blackColor"
        markdown={false}
        markdownStyle={Object {}}
        style={
          Object {
            "color": "rgb(45, 47, 127)",
          }
        }
        type="small"
      >
        Clear
      </Text>
    </Touchable>
  }
  onApply={[Function]}
  onCancel={[Function]}
  title={
    <Text
      accessibilityLabel="Selected: 2 Feb - 8 Feb "
      accessibilityLiveRegion="polite"
      accessibilityTraits={
        Array [
          "header",
        ]
      }
      accessible={true}
      color="blackColor"
      markdown={false}
      markdownStyle={Object {}}
      style={
        Object {
          "alignSelf": "flex-end",
          "color": "rgb(45, 47, 127)",
          "height": 40,
          "paddingTop": 10,
        }
      }
      type="h3"
    >
      2 Feb - 8 Feb
    </Text>
  }
  visible={true}
>
  <DateRangePicker
    dateRange={
      Object {
        "endDate": "2018-02-08",
        "startDate": "2018-02-02",
      }
    }
    forceNewRange={false}
    onChange={[Function]}
  />
</Dialog>
`;

exports[`renders correctly without date selection 1`] = `
<Dialog
  applyButtonLabel="Apply selection"
  applyButtonText="Apply"
  headerRight={
    <Touchable
      accessibilityComponentType="button"
      accessibilityLabel="Clear date selection"
      accessibilityTraits={
        Array [
          "button",
        ]
      }
      accessible={undefined}
      delayLongPress={undefined}
      delayPressIn={50}
      delayPressOut={undefined}
      disabled={undefined}
      hitSlop={undefined}
      onLongPress={undefined}
      onPress={[Function]}
      onPressIn={undefined}
      onPressOut={undefined}
      pressRetentionOffset={undefined}
      style={undefined}
    >
      <Text
        color="blackColor"
        markdown={false}
        markdownStyle={Object {}}
        style={
          Object {
            "color": "rgb(45, 47, 127)",
          }
        }
        type="small"
      >
        Clear
      </Text>
    </Touchable>
  }
  onApply={[Function]}
  onCancel={[Function]}
  title={
    <Text
      accessibilityLabel="Select dates"
      accessibilityLiveRegion="polite"
      accessibilityTraits={
        Array [
          "header",
        ]
      }
      accessible={true}
      color="blackColor"
      markdown={false}
      markdownStyle={Object {}}
      style={
        Object {
          "alignSelf": "flex-end",
          "color": "rgb(45, 47, 127)",
          "height": 40,
          "paddingTop": 10,
        }
      }
      type="h3"
    >
      Select dates
    </Text>
  }
  visible={true}
>
  <DateRangePicker
    dateRange={null}
    forceNewRange={false}
    onChange={[Function]}
  />
</Dialog>
`;
